### PR TITLE
refactor: migration to Jina 3

### PR DESCRIPTION
**Pull request by JohannesMessner**
[DO NOT MERGE] before changing back CI to point to the master branch of workflows-executor once Jina 3 is released

### CODE REVIEW

**Comment by JohannesMessner** [comment]
The test test_delete fails because of an upstream bug; the fix is in the making

File: executor.py [hunk 1]
```diff
@@ -29,6 +33,8 @@ def __init__(
         :param match_args: the arguments to `DocumentArray`'s match function
         :param key_length: the `key_length` keyword argument to
         `DocumentArrayMemmap`'s constructor
+        :param protocol: serialisation protocol for disk access
+        :param compress: compression algorithm for disk access
         :param buffer_pool_size: the `buffer_pool_size` argument to
         `DocumentArrayMemmap`'s constructuor., which stores
             the indexed Documents. During querying, the embeddings of the indexed
```
Comment (nan-wang) [inline executor.py:40]: Please update the docstrings

File: executor.py [hunk 2]
```diff
@@ -147,4 +140,22 @@ def fill_embedding(self, docs: Optional[DocumentArray], **kwargs):
...
+
+    @requests(on='/dump')
+    def dump(self, docs: Optional[DocumentArray]=None, **kwargs):
+        """dump indexed Documents to disk
+        """
+        if docs:
+            self._index.extend(docs)
+        bytes = self._index.to_bytes(protocol=self.protocol, compress=self.compress)
+        with open(os.path.join(self.workspace, SimpleIndexer.FILE_NAME), 'wb') as f:
+            f.write(bytes)
```
Comment (nan-wang) [inline executor.py:153]: What is the purpose of `/dump`? To my understanding, we want to dump the indexed Documents to the disk. If this is the case, we don't need `docs` as inputs.

File: executor.py [hunk 3]
```diff
@@ -147,4 +140,22 @@ def fill_embedding(self, docs: Optional[DocumentArray], **kwargs):
...
+
+    @requests(on='/load')
+    def load_from_disk(self, docs: Optional[DocumentArray]=None, **kwargs):
+        with open(os.path.join(self.workspace, SimpleIndexer.FILE_NAME), 'rb') as f:
+            self._index = DocumentArray.from_bytes(f, protocol=self.protocol, compress=self.compress)
+        if docs:
+            self._index.extend(docs)
```
Comment (nan-wang) [inline executor.py:160]: The same as above. I don't see the rationale of having `docs` as an input argument.

**Review by nan-wang — CHANGES_REQUESTED** [review]

File: executor.py (hunk 2 above)
Comment (JohannesMessner) [reply executor.py:153]: Correct; we misremembered the required parameters for a request-method. Will change

File: executor.py [hunk 4]
```diff
@@ -1,53 +1,52 @@
...
 
+    FILE_NAME = 'index.bin'
+
     def __init__(
         self,
         match_args: Optional[Dict] = None,
-        key_length: int = 64,
-        buffer_pool_size: int = 100_000,
+        protocol: str = 'pickle-array',
+        compress: str = None,
```
Comment (JoanFM) [inline executor.py:26]: fix type hinting, it should be Optional

**Review by JoanFM — CHANGES_REQUESTED** [review]
Please override the `close` method and call `dump` in there

**Review by JoanFM — APPROVED** [review]

**Review by JoanFM — CHANGES_REQUESTED** [review]
Wait

File: executor.py [hunk 5]
```diff
@@ -1,53 +1,50 @@
...
     """
 
+    FILE_NAME = 'index.bin'
+
     def __init__(
         self,
         match_args: Optional[Dict] = None,
-        key_length: int = 64,
-        buffer_pool_size: int = 100_000,
+        protocol: Optional[str] = 'pickle-array',
```
Comment (JoanFM) [inline executor.py]: is this really Optional?

File: executor.py [hunk 6]
```diff
@@ -1,53 +1,50 @@
...
-            the indexed Documents. During querying, the embeddings of the indexed
-            Documents are cached in a buffer pool.
-            `buffer_pool_size` sets the number of Documents to be cached. Make sure
-            it is larger than the total number
-            of indexed Documents to avoid repeating loading embeddings. By default,
-            it is set to `100000`.
-            Check more information at
-            https://docs.jina.ai/api/jina.types.arrays.memmap/?jina.types.arrays
-            .memmap.DocumentArrayMemmap.
+        :param protocol: serialisation protocol for disk access
```
Comment (JoanFM) [inline executor.py:32]: add the possibilities here

**Review by JoanFM — CHANGES_REQUESTED** [review]

**Review by JoanFM — DISMISSED** [review]

**Review by JoanFM — APPROVED** [review]

**Pull request merged**